feat(pkg): Initial implementation

esatterwhite opened the pull request:
Initial implementation of the leader elector

- [x] Kind setup
- [x] Tilt Setup
- [x] Better Tests
- [x] README Docs

esatterwhite committed:
initial commit

Initial project layout and test setup

esatterwhite committed:
feat(elector): initial elector implementation

implements the leadership election process using kubernetes leases

esatterwhite committed:
chore(tilt): adds initial kind and tilt setup for addition testing

includes npm scripts for starting and stoping kind / tilt in addition to
a very simple build to run the elector on kubernetes

esatterwhite committed:
chore(test): include setup chan and initial smoke test

Includes a test to bootstrap a single elector and ensurese that it
claims leadership and shutsdown cleanly

esatterwhite committed:
fix(elector): make sure to close the watcher during shutdown

The watcher maintains a keep alive connection and will hold the process
open indefinately. Store a reference to the abort controller returned by
the watch function and manually stop it during shutdown. If we get an
abort error from the watch callback during the shutdown process, that is
expected

esatterwhite committed:
chore(actions): include kind action to bootstrap cluster

include htlm/kind-action to stand up a cluster to reference and interact
with during tests

darinspivey commented on README.md:
```diff
@@ -1 +1,10 @@
+=======
+## Kubernetes Leader Election
+
+Leader election for kubernetes using leases
```
This could be beefed up.  What's a leader and what do you need a leader for?  Maybe examples?  What problem does this solve?

I'm asking for me too :)

esatterwhite replied:
Yeah, I'll come back through and do that bit last when I'm done fiddling around.
The TLDR is a zero-ish config leader election.

If you are running many pods, and you really want or need one of them to be responsible for preforming some action.

```javascript
if (!elector.current_leader) return
// do important stuff that can only be done by me
```

Think cache manager w/o all the redis coordination and watcher bits.
its pushed off to kubernetes

its a common pattern in kubernetes operators, but all that machinery isn't free unless you're building in Golang and using kubebuilder specifically

esatterwhite committed:
chore(test): adding additional tests for better coverage

includes tests for multiple electors and the case when running without
a kubernetes environment (local dev /testing)

darinspivey reviewed (commented):
Some minor feedback.  I didn't have time to review the tests, but they probably speak for themselves along with the coverage.

esatterwhite commented on lib/elector.js:
```diff
@@ -0,0 +1,412 @@
+      if (!this.#current_leader) {
+        setTimeout(() => {
+          this.#becomeLeader()
+        }, 2000)
```
yes in both cases

esatterwhite replied:
maybe better to debounce

esatterwhite commented on lib/elector.js:
```diff
@@ -0,0 +1,412 @@
+            holderIdentity: this.#LEADER_IDENTITY
+          , leaseDurationSec: this.lease_duration_sec
+          , acquireTime: new V1MicroTime(new Date())
+          , renewTime: new V1MicroTime(new Date())
```
oh, well no, i think this makes more sense. 

> renewTime is a time when the current holder of a lease has last updated the lease

this is probably most accurate

darinspivey replied:
oooh, yes, that's quite a difference.  Time renewed versus time to be renewed.

esatterwhite commented:
:tada: This PR is included in version 1.0.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@codedependant/kubernetes-leader-election/v/1.0.0)
- [GitHub release](https://github.com/esatterwhite/kubernetes-leader-election/releases/tag/v1.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
